File: Lab/src/linkedLists/DoublyLinkedList.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package linkedLists;

/**
 * Reference https://www.programiz.com/dsa/doubly-linked-list for Inserting and
 * removing from position of doubly linked list.
 *
 * @author jackc
 * @param <T>
 */
public class DoublyLinkedList<T> {

    Node<T> head;
    Node<T> tail;
    int count = 0;

    /**
     * Doubly Linked List to string, traverse the list from head to tail. 
     * @return
     */
    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        Node<T> current = head;
        while (current != null) {
            sb.append(current.getValue());

            if (current.getNext() != null) {
                sb.append(", ");
            }
            current = current.getNext();
        }

        return "Linked List: " + sb;
    }

    /**
     * Doubly Linked List to string in traverse the doubly linked list from tail to head. 
     * @return
     */
    public String toStringReverse() {
        StringBuilder sb = new StringBuilder();
        Node<T> current = tail;
        while (current != null) {
            sb.append(current.getValue());

            if (current.getPrevious() != null) {
                sb.append(", ");
            }
            current = current.getPrevious();
        }

        return "Linked List: " + sb;
    }

    /**
     * Search for item in doubly linked list. 
     * @param item
     * @return
     */
    public boolean contains(T item) {
        Node<T> current = head;
        while (current != null) {
            if (current.value.equals(item)) {
                return true;
            }

            current = current.next;

        }
        return false;
    }
    //END     public boolean contains(T item)

    /**
     * Add a node to the start of the list. 
     * @param node
     */
    public void addFirst(Node<T> node) {
        Node<T> temp = head;
        head = node;
        head.next = temp;
        if (count == 0) {
            //if the liist is empty, the head and tail 
            //point to the new node: 
            tail = head;
        } else {
            /*
            Before: head → 5 ←→ 7 → null
            After: head → 3 ←→ 5 ←→ 7 → null
            temp.previous was null and now head
             */
            temp.previous = head;
        }
        count++;
    }
    //END     public void addFirst(Node<T> node)

    /**
     * Add a node to the end of the list. 
     * @param node
     */
    public void addLast(Node<T> node) {
        if (count == 0) // no nodes
        {
            head = node;
        } else {
            tail.next = node;
            //Before: Head → 3 ←→ 5 → null
            //After: Head → 3 ←→ 5 ←→ 7 → null
            node.previous = tail;
        }
        tail = node;
        count++;
    }
    //END     public void addLast(Node<T> node)

    /**
     * Insert node at the position specified in the doubly linked list. 
     * @param position
     * @param newNode
     */
    public void insertMiddle(int position, Node<T> newNode) {
        //Save the head
        Node<T> temp = head;

        if (position < 1) {
            System.out.println("Position can't be 0 or less.");
        }

        for (int i = 2; i < position; i++) {
            if (temp.next != null) {
                temp = temp.next;
            }
        }

        newNode.next = temp.next;

        //Set the nodes prvious to the current node
        newNode.previous = temp;

        //Don't foreget to set the next node's head to point back to the new node. Forgot first time. 
        newNode.next.previous = newNode;
        temp.next = newNode;

    }

    /**
     * Remove node at the position specified in the doubly linked list. 
     * @param postion
     */
    public void removeMiddle(int postion) {
        Node<T> temp = head;

        for (int i = 2; i < postion; i++) {
            if (temp.next != null) {
                temp = temp.next;
            }
        }
        temp.next = temp.next.next;
        temp.next.previous = temp;
    }

    /**
     * Remove node from head of list. 
     */
    public void removeFirst() {
        if (count != 0) {
            //Bfore: Head: → 3 ←→ 5
            //After: Head: → 5

            //Head → → null
            //Head → null
            head = head.next;
            count--;
            if (count == 0) {
                tail = null;
            } else {
                //5.previous was 3, now 5.prev is null
                head.previous = null;
            }
        }
    }
    //END     public void removeFirst()

    /**
     * Remove node from tail of list. 
     */
    public void removeLast() {
        if (count != 0) {
            if (count == 1) {
                head = null;
                tail = null;
            } else {
                tail.previous.next = null;
                tail = tail.previous;
            }
            count--;
        }
    }
    //END     public void removeLast() 

    //GETTERS AND SETTERS

    /**
     *
     * @return
     */
    public Node<T> getHead() {
        return head;
    }

    /**
     *
     * @param head
     */
    public void setHead(Node<T> head) {
        this.head = head;
    }

    /**
     *
     * @return
     */
    public Node<T> getTail() {
        return tail;
    }

    /**
     *
     * @param tail
     */
    public void setTail(Node<T> tail) {
        this.tail = tail;
    }

    /**
     *
     * @return
     */
    public int getCount() {
        return count;
    }

    /**
     *
     * @param count
     */
    public void setCount(int count) {
        this.count = count;
    }

}

File: CSharp-Code/Linked-Lists/AssessableLab/DoublyLinkedList.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AssessableLab
{
    class DoublyLinkedList<T> : System.Collections.Generic.ICollection<T>
    {
        //CONSTRUCTOR
        public LinkedListNode<T> Head
        {
            get;
            private set;
        }

        //CONSTRUCTOR
        public LinkedListNode<T> Tail
        {
            get;
            private set;
        }

        //Overload AddFirst()
        public void AddFirst(T value)
        {
            AddFirst(new LinkedListNode<T>(value));
        }
        public void AddFirst(LinkedListNode<T> node)
        {
            //Save head node so we don't lose it
            LinkedListNode<T> temp = Head;

            //Point head to new node
            Head = node;

            //Insert est of the list behind the head
            Head.Next = temp;

            //IF list was empty then Head and Tail point to the new node
            if (Count == 0)
                Tail = Head;
            //ELSE 
            //Before: Head ------> 5 <-> 7 -> null
            //After: Head -> 3 <-> 5 <-> 7 -> null
            else
                temp.Previous = Head;

            Count++;
        }


        //Overload AddLast()
        public void AddLast(T value)
        {
            AddLast(new LinkedListNode<T>(value));
        }
        public void AddLast(LinkedListNode<T> node)
        {
            if (Count == 0)
            {
                Head = node;
            }
            else
            {
                Tail.Next = node;
                //Before: Head -> 3 <-> 5 -> null
                //After:  Head ->3 <-> 5 <-> 7 -> null
                // Tail was 5, now is 7
                node.Previous = Tail;
            }
            Tail = node;
            Count++;
        }

        //RemoveFirst()
        public void RemoveFirst()
        {
            if (Count != 0)
            {
                //Before: Head -> 3 <-> 5 
                //After:  Head -------> 5
                //or 
                //Before: Head -> 3 -> null
                //After: Head -------> null
                Head = Head.Next;

                Count--;

                if (Count == 0)
                    Tail = null;
                else
                    //5.Previous was 3, now 5.previous is now null
                    Head.Previous = null;
            }
        }

        //RemoveLast()
        public void RemoveLast()
        {
            if (Count != 0)
            {
                if (Count == 1)
                {
                    Head = null;
                    Tail = null;
                }
                else
                {
                    //Before: Head -> 3 <-> 5 <-> 7
                    //     Tail: 7
                    //After:  Head ->3 <-> 5 -> null
                    //      Tail: 5
                    Tail.Previous.Next = null;
                    Tail = Tail.Previous;
                }

                Count--;
            }
        }

        //Count property
        public int Count
        {
            get;
            private set;
        }

        //Add()
        public void Add(T item)
        {
            AddFirst(item);
        }

        //Contains()
        public bool Contains(T item)
        {
            LinkedListNode<T> current = Head;
            while (current != null)
            {
                //Head -> 3 <-> 5 <-> 7
                //Value : 5
                if (current.Value.Equals(item))
                    return true;

                current = current.Next;
            }
            return false;
        }

        //CopyTo(T[] array, int arrayInded)
        public void CopyTo(T[] array, int arrayIndex)
        {
            LinkedListNode<T> current = Head;
            while (current != null)
            {
                array[arrayIndex++] = current.Value;
                current = current.Next;
            }
        }

        //IsReadOnly()
        public bool IsReadOnly
        {
            get { return false; }
        }

        //Remove metod
        public bool Remove(T item)
        {
            LinkedListNode<T> previous = null;
            LinkedListNode<T> current = Head;

            /* 1. Empty list, do nothing 
             * 2. Single node: (previous is null). 
             * 3 Many nodes:
             *      a. Node to remove is the first node. 
             *      b. Node to remove is the middle or last node. */

            //While NOT case 1?
            while (current != null)
            {
                //Before Head -> 3 <-> 5 <-> 7 -> null
                //After  Head -> 3 <-------> 7 -> null

                if (current.Value.Equals(item))
                {
                    //It's a node in the middle or end: 
                    if (previous != null)
                    {
                        //Case 3b:
                        previous.Next = current.Next;

                        //it was the end os update Tail
                        if (current.Next == null)
                        {
                            Tail = previous;
                        }
                        else
                        {
                            /*
                             * Before Head -> 3 <-> 5 <-> 7 -> null 
                             * After  Head -> 3 <-------> 7 -> null 
                             * Previous = 3
                             * Current = 5
                             * current.Next = 7
                             * So...... 7.Previous = 3
                              */
                            current.Next.Previous = previous;
                        }

                        Count--;
                    }
                    else
                    {
                        //CAse 2 or 3a
                        RemoveFirst();
                    }

                    return true;
                }

                //Case 1?
                previous = current;
                current = current.Next;
            }

            return false;
        }
        //END method public bool Remove(T item)



        // Implement the TWO GetEnumerator ( ) to enumerate over the linked list values from Head to Tail 
        System.Collections.Generic.IEnumerator<T> System.Collections.Generic.IEnumerable<T>.GetEnumerator()
        {
            LinkedListNode<T> current = Head;
            while (current != null)
            {
                yield return current.Value;
                current = current.Next;
            }
        }
        System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
        {
            return ((System.Collections.Generic.IEnumerable<T>)this).GetEnumerator();
        }

        //Clear()
        public void Clear()
        {
            Head = null;
            Tail = null;
            Count = 0;
        }
    }
}
